src/groth16/aggregate/accumulator.rs

```diff
@@ -2,13 +2,133 @@
 use crate::bls::Engine;
 #[cfg(feature = "blst")]
 use blstrs::PairingCurveAffine;
-use core::default::Default;
+
+use crossbeam_channel::{bounded, Receiver, Sender};
 use ff::{Field, PrimeField};
 use groupy::{CurveAffine, CurveProjective};
 #[cfg(feature = "pairing")]
 use paired::{Engine, PairingCurveAffine};
-use rand::rngs::OsRng;
 use rayon::prelude::*;
+
+use std::sync::{
+    atomic::{AtomicBool, Ordering::SeqCst},
+    Arc, Mutex,
+};
+
+pub struct PairingChecks<E: Engine, R: rand::RngCore + Send> {
+    valid: Arc<AtomicBool>,
+    merge_send: Sender<PairingCheck<E>>,
+    valid_recv: Receiver<bool>,
+    rng: Mutex<R>,
+    /// Ensures that the non randomized check is only added exactly once.
+    non_random_check_done: AtomicBool,
+}
+
+impl<E: Engine, R: rand::RngCore + Send> PairingChecks<E, R> {
+    pub fn new(rng: R) -> Self {
+        let (merge_send, merge_recv): (Sender<PairingCheck<E>>, Receiver<PairingCheck<E>>) =
+            bounded(10);
+        let (valid_send, valid_recv) = bounded(1);
+
+        let valid = Arc::new(AtomicBool::new(true));
+        let valid_copy = valid.clone();
+
+        rayon::spawn(move || {
+            let mut acc = PairingCheck::new();
+            while let Ok(tuple) = merge_recv.recv() {
+                // only do work as long as we know we are still valid
+                if valid_copy.load(SeqCst) {
+                    acc.merge(&tuple);
+                }
+            }
+            if valid_copy.load(SeqCst) {
+                valid_send.send(acc.verify()).expect("failed to send");
+            }
+        });
+
+        PairingChecks {
+            valid,
+            merge_send,
+            valid_recv,
+            rng: Mutex::new(rng),
+            non_random_check_done: AtomicBool::new(false),
+        }
+    }
+
+    /// Fails the whole check.
+    pub fn invalidate(&self) {
+        self.valid.store(false, SeqCst);
+    }
+
+    fn merge_pair(&self, result: E::Fqk, exp: E::Fqk) {
+        self.merge(PairingCheck::from_pair(result, exp));
+    }
+
+    fn merge_miller_one(&self, result: E::Fqk) {
+        self.merge(PairingCheck::from_miller_one(result));
+    }
+
+    pub fn merge_parts<F, G, H, I>(&self, f: F, g: G, h: H, i: I)
+    where
+        F: Fn() -> E::Fqk + Send,
+        G: Fn() -> E::Fqk + Send,
+        H: Fn() -> E::Fqk + Send,
+        I: Fn() -> (E::Fqk, E::Fqk) + Send,
+    {
+        assert!(
+            !self.non_random_check_done.load(SeqCst),
+            "merge_parts can only be called once"
+        );
+
+        rayon::scope(|s| {
+            s.spawn(move |_| {
+                self.merge_miller_one(f());
+            });
+            s.spawn(move |_| {
+                self.merge_miller_one(g());
+            });
+            s.spawn(move |_| {
+                self.merge_miller_one(h());
+            });
+            s.spawn(move |_| {
+                let (a, b) = i();
+                self.merge_pair(a, b);
```

Comment:
This is a single value when you use it - why asking for a tuple ?

Reply:
I just kept it inline with the previous usage of `from_pair`, I agree it makes sense to streamline